framework/include/base/Assembly.h

```diff
@@ -2800,63 +2861,52 @@ void
+#endif
+                             ,
                              LocalFunctor &
 #ifndef MOOSE_GLOBAL_AD_INDEXING
                                  local_functor
 #endif
 )
 {
 #ifdef MOOSE_GLOBAL_AD_INDEXING
-  // Need to make a copy because we might modify this in constrain_element_matrix
-  std::vector<dof_id_type> row_indices = input_row_indices;
-
-  mooseAssert(residuals.size() == row_indices.size(),
-              "The number of residuals should match the number of dof indices");
-  mooseAssert(residuals.size() >= 1, "Why you calling me with no residuals?");
-
-  const auto & compare_dofs = residuals[0].derivatives().nude_indices();
-#ifndef NDEBUG
-  auto compare_dofs_set = std::set<dof_id_type>(compare_dofs.begin(), compare_dofs.end());
-
-  for (auto resid_it = residuals.begin() + 1; resid_it != residuals.end(); ++resid_it)
-  {
-    auto current_dofs_set = std::set<dof_id_type>(resid_it->derivatives().nude_indices().begin(),
-                                                  resid_it->derivatives().nude_indices().end());
-    mooseAssert(compare_dofs_set == current_dofs_set,
-                "We're going to see whether the dof sets are the same. IIRC the degree of freedom "
-                "dependence (as indicated by the dof index set held by the ADReal) has to be the "
-                "same for every residual passed to this method otherwise constrain_element_matrix "
-                "will not work.");
-  }
+  processResiduals(residuals, input_row_indices, {}, matrix_tags, scaling_factor);
+#else
+  local_functor(residuals, input_row_indices, matrix_tags);
 #endif
-  auto column_indices = std::vector<dof_id_type>(compare_dofs.begin(), compare_dofs.end());
+}
 
-  // If there's no derivatives then there is nothing to do. Moreover, if we pass zero size column
-  // indices to constrain_element_matrix then we will potentially get errors out of BLAS
-  if (!column_indices.size())
+template <typename T>
+void
+Assembly::processResiduals(const std::vector<T> & residuals,
+                           const std::vector<dof_id_type> & input_row_indices,
+                           const std::set<TagID> & vector_tags,
+                           const Real scaling_factor)
+{
+  if (!computingResidual() || vector_tags.empty())
     return;
 
-  DenseMatrix<Number> element_matrix(row_indices.size(), column_indices.size());
-  for (const auto i : index_range(row_indices))
-  {
-    const auto row_index = row_indices[i];
-    const Real scalar = _scaling_vector ? (*_scaling_vector)(row_index) : 1.;
+  mooseAssert(residuals.size() == input_row_indices.size(),
+              "The number of residuals should match the number of dof indices");
+  mooseAssert(residuals.size() >= 1, "Why you calling me with no residuals?");
 
-    const auto & sparse_derivatives = residuals[i].derivatives();
+  // Need to make a copy because we might modify this in constrain_element_vector
+  std::vector<dof_id_type> row_indices = input_row_indices;
```

Comment:
Because we might modify the vector in `constrain_element_vector` and the input indices correspond to the unconstrained dof indices, which is data we don't want to modify. Let me know if I can make the comment clearer